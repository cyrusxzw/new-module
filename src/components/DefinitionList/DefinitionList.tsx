import React, { Fragment, forwardRef } from 'react';
import cx from 'classnames';
import { isObjectPopulatedArray } from '~/utils/objects';
import styles from './DefinitionList.module.css';
import type { DefinitionListProps } from './DefinitionList.types';

const DefinitionList = forwardRef<HTMLDListElement, DefinitionListProps>(
  function DefinitionListRef(
    {
      className,
      hasBottomBorder = false,
      isVisible = true,
      items = [],
      theme = 'dark',
    },
    ref,
  ) {
    if (!isObjectPopulatedArray(items)) {
      return null;
    }

    const classSet = cx(
      styles.base,
      styles[theme],
      { [styles.hasBottomBorder]: hasBottomBorder },
      className,
    );
    const termClassSet = cx(styles.term, { [styles.slideIn]: isVisible });
    const descriptionClassSet = cx(styles.description, {
      [styles.slideIn]: isVisible,
    });

    return (
      <dl
        className={classSet}
        data-testid="data-testid-DefinitionList"
        ref={ref}
      >
        {items
          .filter(({ description, term }) => description || term)
          .map(({ description, id, term }) => (
            <Fragment key={id}>
              <dt className={termClassSet}>{term}</dt>
              <dd className={descriptionClassSet}>{description}</dd>
            </Fragment>
          ))}
      </dl>
    );
  },
);

export { DefinitionList };
