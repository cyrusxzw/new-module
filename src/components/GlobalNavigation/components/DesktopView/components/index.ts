export { CloseButton } from './CloseButton';
export { CollectionImage } from './CollectionImage';
export { CollectionLayout } from './CollectionLayout';
export { CollectionList } from './CollectionList';
export { MenuItem } from './MenuItem';
export { Panel } from './Panel';
export { PrimaryMenu } from './PrimaryMenu';
export { SecondaryMenu } from './SecondaryMenu';
