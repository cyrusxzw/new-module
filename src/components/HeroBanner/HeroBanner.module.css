@import 'styles/index';

.base {
  position: relative;
  min-height: 80vh;

  @media (--viewport-lg) {
    display: flex;
    min-height: 65vh;
    flex-direction: row;
  }
}

.reverse {
  @media (--viewport-lg) {
    flex-direction: row-reverse;
  }
}

.topOffest {
  padding-top: 40px;
  margin-top: -120px;

  @media (--viewport-md) {
    padding-top: 50px;
    margin-top: -130px;
  }

  @media (--viewport-lg) {
    padding-top: 75px;
    margin-top: -155px;
  }
}
