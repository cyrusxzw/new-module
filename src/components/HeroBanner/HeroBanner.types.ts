import { ReactElement, ReactNode } from 'react';
import type { Themes } from '~/types';

type AnchorPoints = 'center' | 'left' | 'right';

type HeroBannerProps = {
  backgroundColor?: string;
  className?: string;
  containMedia?: AnchorPoints;
  content?: ReactNode;
  copy?: {
    description?: string;
    eyebrow?: string;
    heading?: string;
    subHeading?: string;
  };
  foregroundImage?: ReactElement;
  foregroundImageLink?: {
    url?: string;
    title?: string;
  };
  hasFullWidthMedia?: boolean;
  hasSerifFontHeading?: boolean;
  hasTopOffset?: boolean;
  isReverse?: boolean;
  media: ReactElement;
  theme?: Themes;
};

export type { HeroBannerProps };
