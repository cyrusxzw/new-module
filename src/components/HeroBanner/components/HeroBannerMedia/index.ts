import { HeroBannerMedia } from './HeroBannerMedia';

export { HeroBannerMedia };
