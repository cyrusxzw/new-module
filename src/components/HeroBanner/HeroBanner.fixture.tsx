import React from 'react';
import { Hyperlink } from '~/components/Hyperlink';
import { HYPERLINK_STYLE_TYPES } from '~/constants';

const { EXTERNAL_NO_ICON_TEXT_LINK } = HYPERLINK_STYLE_TYPES;

const HeroBannerFixture = {
  copy: {
    default: {
      eyebrow: 'Praesent tincidunt accumsan',
      heading: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
      subHeading: '',
      description:
        'Aliquam turpis libero, posuere sed nisl in, porta facilisis lorem. Phasellus eu turpis in quam efficitur hendrerit.',
    },
    wideHeader: {
      eyebrow: 'Quisque blandit est erat',
      heading: 'Fusce et dui nisi',
      description: (
        <p>
          Aliquam lectus purus, gravida id metus dignissim, efficitur tempus
          lacus. Maecenas a cursus diam. Vivamus ultrices nisi non ex placerat,
          vel elementum sapien condimentum. Lorem ipsum dolor sit amet,
          consectetur adipiscing elit, efficitur tempus lacus. Maecenas a cursus
          diam. Vivamus ultrices nisi non ex placerat, vel elementum sapien
          condimentum.
        </p>
      ),
    },
  },
  image: {
    default: {
      altText: 'alt image',
      large:
        './assets/images/HeroBanner/Aesop_Lunar_New_Year_2021_Homepage_Primary_Full_Bleed_Desktop_2880x1044px.jpg',
      medium:
        './assets/images/HeroBanner/Aesop_Lunar_New_Year_2021_Homepage_Primary_Full_Bleed_Tablet_1536x1730px.jpg',
      small:
        './assets/images/HeroBanner/Aesop_Lunar_New_Year_2021_Homepage_Primary_Full_Bleed_Mobile_640x640px.jpg',
    },
    fiftyFifty: {
      altText: 'alt image',
      large:
        './assets/images/HeroBanner/Aesop_Redemption_Body_Scrub_2021_Homepage_Primary_50-50_Desktop_1440x1500px.jpg',
      medium:
        './assets/images/HeroBanner/Aesop_Redemption_Body_Scrub_2021_Homepage_Primary_50-50_Tablet_1536x950px.jpg',
      small:
        './assets/images/HeroBanner/Aesop_Redemption_Body_Scrub_2021_Homepage_Primary_50-50_Mobile_640x640px.jpg',
    },
    articleHeader: {
      altText: 'alt image',
      large:
        './assets/images/HeroBanner/Aesop_A_Guide_to_Caring_for_Oily_Skin_Web_Primary_Full_Bleed_Desktop_2880x580px.jpg',
      medium:
        './assets/images/HeroBanner/Aesop_A_Guide_to_Caring_for_Oily_Skin_Web_Primary_Full_Bleed_Tablet_1536x520px.jpg',
      small:
        './assets/images/HeroBanner/Aesop_A_Guide_to_Caring_for_Oily_Skin_Web_Primary_Full_Bleed_Mobile_640x640px.jpg',
    },
    wideHeader: {
      altText: 'alt image',
      large: './assets/images/HeroBanner/wide-header-large.webp',
      medium: './assets/images/HeroBanner/wide-header-medium.webp',
      small: './assets/images/HeroBanner/wide-header-small.webp',
    },
    foreground: {
      altText: 'alt image',
      large: './assets/images/MediaWithContent/foreground-large.jpg',
      medium: './assets/images/MediaWithContent/foreground-large.jpg',
      small: './assets/images/MediaWithContent/foreground-large.jpg',
    },
  },
  linkButtonGroup: {
    textAlign: 'left',
    hasFitContent: true,
    children: (
      <>
        <Hyperlink style="Internal Button Link" url="/">
          Proin quisque
        </Hyperlink>
        <Hyperlink style="Internal Button Link" url="/">
          Quisque
        </Hyperlink>
      </>
    ),
  },
  subNav: {
    links: [
      {
        id: 'aenean-magna-lectus',
        style: EXTERNAL_NO_ICON_TEXT_LINK,
        children: 'Aenean magna lectus',
        url: 'https://www.aesop.com',
      },
      {
        id: 'phasellus-magna-nulla',
        style: EXTERNAL_NO_ICON_TEXT_LINK,
        children: 'Phasellus magna nulla',
        url: 'https://www.aesop.com',
      },
      {
        id: 'sed-facilisis-felis-venenatis',
        style: EXTERNAL_NO_ICON_TEXT_LINK,
        children: 'Sed facilisis felis venenatis',
        url: 'https://www.aesop.com',
      },
      {
        id: 'curabitur-auctor',
        style: EXTERNAL_NO_ICON_TEXT_LINK,
        children: 'Curabitur auctor',
        url: 'https://www.aesop.com',
      },
    ],
  },
  video: {
    source:
      'https://player.vimeo.com/external/462935380.hd.mp4?s=01e048a4b5942b2c6648ffe9965ac5c79ebfc545&profile_id=169',
  },
};

export { HeroBannerFixture };
