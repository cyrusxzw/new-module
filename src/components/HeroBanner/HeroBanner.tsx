import React, { forwardRef, FC } from 'react';
import cx from 'classnames';
import { HeroBannerMedia } from './components/HeroBannerMedia';
import { HeroBannerContent } from './components/HeroBannerContent';
import type { HeroBannerProps } from './HeroBanner.types';
import styles from './HeroBanner.module.css';

const HeroBanner: FC<HeroBannerProps> = forwardRef(function HeroBannerRef(
  {
    backgroundColor = '#F6F5E8',
    className,
    containMedia,
    content,
    copy,
    foregroundImage,
    foregroundImageLink,
    hasFullWidthMedia = false,
    hasSerifFontHeading = true,
    hasTopOffset = false,
    isReverse = false,
    media,
    theme = 'dark',
  },
  ref: React.LegacyRef<HTMLImageElement>,
) {
  const classSet = cx(
    styles.base,
    { [styles.reverse]: isReverse },
    { [styles.topOffest]: hasTopOffset },
    className,
  );

  const style = backgroundColor ? { backgroundColor } : undefined;

  return (
    <div className={classSet} ref={ref} style={style}>
      <HeroBannerMedia
        containMedia={containMedia}
        foregroundImage={foregroundImage}
        foregroundImageLink={foregroundImageLink}
        hasFullWidthMedia={hasFullWidthMedia}
        media={media}
      />
      <HeroBannerContent
        content={content}
        copy={copy}
        hasFullWidthMedia={hasFullWidthMedia}
        hasSerifFontHeading={hasSerifFontHeading}
        hasTopOffset={hasTopOffset}
        isReverse={isReverse}
        theme={theme}
      />
    </div>
  );
});

export { HeroBanner };
