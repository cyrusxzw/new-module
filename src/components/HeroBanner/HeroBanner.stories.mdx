import { Fragment } from 'react';
import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { GridOverlay } from '~.storybook/GridOverlay.js';
import { boolean, select, text, radios } from '@storybook/addon-knobs';
import { DefinitionList } from '~/components/DefinitionList';
import { Hyperlink } from '~/components/Hyperlink';
import { Image } from '~/components/Image';
import { KitList } from '~/components/KitList';
import { LinkButtonGroup } from '~/components/LinkButtonGroup';
import { SubNav } from '~/components/SubNav';
import { Video } from '~/components/Video';
import { HeroBanner } from './HeroBanner';
import { HeroBannerFixture } from './HeroBanner.fixture';

<Meta title="Components/HeroBanner" component={HeroBanner} />

# Hero Banner

### Hero Image

<Canvas>
  <Story name="Hero Image">
    <HeroBanner
      backgroundColor="#633e43"
      content={(() => {
        const options = {
          SubNav: {
            Component: SubNav,
            props: HeroBannerFixture.subNav,
          },
          LinkButtonGroup: {
            Component: LinkButtonGroup,
            props: HeroBannerFixture.linkButtonGroup,
          },
        };
        const componentName = select(
          'content',
          Object.keys(options),
          'LinkButtonGroup',
        );
        const { Component, props } = options[componentName];
        return <Component {...props} theme="light" />;
      })()}
      copy={HeroBannerFixture.copy.expectations}
      hasFullWidthMedia={false}
      isReverse={true}
      media={
        <Image
          altText={HeroBannerFixture.image.hero.altText}
          large={HeroBannerFixture.image.hero.large}
          medium={HeroBannerFixture.image.hero.medium}
          small={HeroBannerFixture.image.hero.small}
        />
      }
      theme="light"
    />
  </Story>
</Canvas>

### Hero Image Full Width

<Canvas>
  <Story name="Hero Image Full Width">
    <HeroBanner
      backgroundColor="#222e56"
      content={(() => {
        const options = {
          SubNav: {
            Component: SubNav,
            props: HeroBannerFixture.subNav,
          },
          LinkButtonGroup: {
            Component: LinkButtonGroup,
            props: HeroBannerFixture.linkButtonGroup,
          },
        };
        const componentName = select(
          'content',
          Object.keys(options),
          'LinkButtonGroup',
        );
        const { Component, props } = options[componentName];
        return <Component {...props} theme="light" />;
      })()}
      copy={HeroBannerFixture.copy.default}
      hasFullWidthMedia={true}
      isReverse={true}
      media={
        <Image
          altText={HeroBannerFixture.image.expectations.altText}
          large={HeroBannerFixture.image.expectations.large}
          medium={HeroBannerFixture.image.expectations.medium}
          small={HeroBannerFixture.image.expectations.small}
        />
      }
      theme="light"
    />
  </Story>
</Canvas>

### Hero Image with Foreground Image

<Canvas>
  <Story name="Hero Image with Foreground Image">
    <HeroBanner
      backgroundColor="#f5f3e6"
      content={<Hyperlink {...HeroBannerFixture.hyperlinkButtonDark} />}
      copy={HeroBannerFixture.copy.expectations}
      foregroundImage={
        <Image
          altText={HeroBannerFixture.image.foreground.altText}
          large={HeroBannerFixture.image.foreground.large}
          medium={HeroBannerFixture.image.foreground.medium}
          small={HeroBannerFixture.image.foreground.small}
        />
      }
      foregroundImageLink={HeroBannerFixture.foregroundImageLink}
      hasFullWidthMedia={false}
      isReverse={true}
      media={
        <Image
          altText={HeroBannerFixture.image.background.altText}
          large={HeroBannerFixture.image.background.large}
          medium={HeroBannerFixture.image.background.medium}
          small={HeroBannerFixture.image.background.small}
        />
      }
      theme="dark"
    />
  </Story>
</Canvas>

### Hero Video

<Canvas>
  <Story name="Hero Video">
    <HeroBanner
      backgroundColor="#633e43"
      content={
        <Fragment>
          <LinkButtonGroup
            {...HeroBannerFixture.linkButtonGroup}
            theme="light"
          />
        </Fragment>
      }
      copy={HeroBannerFixture.copy.expectations}
      hasFullWidthMedia={false}
      isReverse={true}
      media={
        <Video
          fallbackImage={HeroBannerFixture.video.fallbackImage}
          hasAllowAudio={false}
          hasAutoplay={true}
          hasLoop={true}
          hasPlayInFullScreen={false}
          hasControls={false}
          isFullWidth={false}
          large={HeroBannerFixture.video.source}
          medium={HeroBannerFixture.video.source}
          small={HeroBannerFixture.video.source}
        />
      }
      theme="light"
    />
  </Story>
</Canvas>

### Hero Video full width

<Canvas>
  <Story name="Hero Video full width">
    <HeroBanner
      backgroundColor="#000"
      copy={HeroBannerFixture.copy.default}
      hasFullWidthMedia={true}
      isReverse={true}
      media={
        <Video
          fallbackImage={HeroBannerFixture.video.fallbackImage}
          hasAllowAudio={false}
          hasAutoplay={true}
          hasLoop={true}
          hasPlayInFullScreen={false}
          hasControls={false}
          isFullWidth={false}
          isHeroFullWidth={true}
          large={HeroBannerFixture.video.source}
          medium={HeroBannerFixture.video.source}
          small={HeroBannerFixture.video.source}
        />
      }
      theme="light"
    />
  </Story>
</Canvas>

### Hero Full width with Foreground Image

<Canvas>
  <Story name="Hero Full width with Foreground Image">
    <HeroBanner
      backgroundColor="#f5f3e6"
      content={<Hyperlink {...HeroBannerFixture.hyperlinkButtonDark} />}
      copy={HeroBannerFixture.copy.expectations}
      foregroundImage={
        <Image
          altText={HeroBannerFixture.image.foreground.altText}
          large={HeroBannerFixture.image.foreground.large}
          medium={HeroBannerFixture.image.foreground.medium}
          small={HeroBannerFixture.image.foreground.small}
        />
      }
      foregroundImageLink={HeroBannerFixture.foregroundImageLink}
      hasFullWidthMedia={true}
      isReverse={true}
      media={
        <Image
          altText={HeroBannerFixture.image.background.altText}
          large={HeroBannerFixture.image.background.large}
          medium={HeroBannerFixture.image.background.medium}
          small={HeroBannerFixture.image.background.small}
        />
      }
      theme="dark"
    />
  </Story>
</Canvas>
