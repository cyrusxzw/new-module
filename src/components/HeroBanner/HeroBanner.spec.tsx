import React from 'react';
import { render } from '@testing-library/react';
import { DefinitionList } from '~/components/DefinitionList';
import { Image } from '~/components/Image';
import { HeroBanner } from './HeroBanner';
import { HeroBannerFixture } from './HeroBanner.fixture';

describe('<HeroBanner />', () => {
  it('should be defined', () => {
    expect(HeroBanner).toBeDefined();
  });

  it('should render base component correctly', () => {
    const { container } = render(
      <HeroBanner
        content={
          <DefinitionList items={HeroBannerFixture.definitionList.items} />
        }
        copy={HeroBannerFixture.copy.expectations}
        media={<Image {...HeroBannerFixture.image.default} />}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
