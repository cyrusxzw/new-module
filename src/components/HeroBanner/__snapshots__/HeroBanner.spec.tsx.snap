// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroBanner /> should render base component correctly 1`] = `
<div>
  <div
    class="base"
    style="background-color: rgb(246, 245, 232);"
  >
    <div
      class="base undefined"
    >
      <figure
        class="base dark figure"
      >
        <picture
          class="base picture dark"
        >
          <source
            media="(min-width: 1025px)"
            srcset="https://www.aesop.com/medias/Aesop-Skin-Classic-Purifying-Facial-Exfoliant-Paste-Texture-50-50-Mobile-640x640px.jpg?context=bWFzdGVyfGltYWdlc3wyNDg1OHxpbWFnZS9qcGVnfGltYWdlcy9oMjIvaDBkLzkxMTYyMjg1MTc5MTguanBnfDJjZDUzMjI4N2EzZDgyYzE1ZTk3MTM4OWFjODU5N2Q2NjQ4YTAxNzcwZmE4YjBlZDQ2YTJmYTRhOTI5ZWQyY2I"
          />
          <source
            media="(min-width: 640px)"
            srcset="https://www.aesop.com/medias/Aesop-Skin-Classic-Purifying-Facial-Exfoliant-Paste-Texture-50-50-Mobile-640x640px.jpg?context=bWFzdGVyfGltYWdlc3wyNDg1OHxpbWFnZS9qcGVnfGltYWdlcy9oMjIvaDBkLzkxMTYyMjg1MTc5MTguanBnfDJjZDUzMjI4N2EzZDgyYzE1ZTk3MTM4OWFjODU5N2Q2NjQ4YTAxNzcwZmE4YjBlZDQ2YTJmYTRhOTI5ZWQyY2I"
          />
          <source
            media="(min-width: 0px)"
            srcset="https://www.aesop.com/medias/Aesop-Skin-Classic-Purifying-Facial-Exfoliant-Paste-Texture-50-50-Mobile-640x640px.jpg?context=bWFzdGVyfGltYWdlc3wyNDg1OHxpbWFnZS9qcGVnfGltYWdlcy9oMjIvaDBkLzkxMTYyMjg1MTc5MTguanBnfDJjZDUzMjI4N2EzZDgyYzE1ZTk3MTM4OWFjODU5N2Q2NjQ4YTAxNzcwZmE4YjBlZDQ2YTJmYTRhOTI5ZWQyY2I"
          />
          <img
            alt="alt image"
            loading="lazy"
          />
        </picture>
      </figure>
    </div>
    <div
      class="base dark"
    >
      <div
        class="innerWrapper"
      >
        <header
          class="header base"
        >
          <h1
            class="base serifFont xLarge dark heading"
          >
            Fusce et dui nisi
          </h1>
          <h2
            class="base mediumWeightFont xXSmall dark eyebrow"
          >
            Quisque blandit est erat
          </h2>
        </header>
        <div
          class="content"
        >
          <div
            class="set dark large description"
          >
            Aliquam lectus purus, gravida id metus dignissim, efficitur tempus lacus. Maecenas a cursus diam. Vivamus ultrices nisi non ex placerat, vel elementum sapien condimentum.
          </div>
          <dl
            class="base dark"
            data-testid="data-testid-DefinitionList"
          >
            <dt
              class="term slideIn"
            >
              Aliquam
            </dt>
            <dd
              class="description slideIn"
            >
              Proin vulputate eros quis lectus pretium gravida.
            </dd>
            <dt
              class="term slideIn"
            >
              Etiam eu augue quam
            </dt>
            <dd
              class="description slideIn"
            >
              Pulvinar nulla enim, eu elementum lacus vehicula sed.
            </dd>
            <dt
              class="term slideIn"
            >
              Duis at elit
            </dt>
            <dd
              class="description slideIn"
            >
              Sed ac ipsum vel risus aliquam rutrum.
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</div>
`;
